File: blastcircuit.cs
// Copyright 2025 William Stafford Parsons

public class BlastCircuit {
  public ulong a;
  public ulong b;
  public ulong c;
  public ulong d;

  public ulong Next() {
    ulong mix = a ^ b;

    a += 111111111111111;
    b = (b >> 3) + c;
    c = d;
    d = ((d << 21) | (d >> 43)) + mix;
    return mix;
  }
}

File: blastcircuit.c
// Copyright 2025 William Stafford Parsons

#include <stdint.h>

struct blastcircuit_state {
  uint64_t a;
  uint64_t b;
  uint64_t c;
  uint64_t d;
};

uint64_t blastcircuit(struct blastcircuit_state *s) {
  uint64_t mix = s->a ^ s->b;

  s->a += 111111111111111;
  s->b = (s->b >> 3) + s->c;
  s->c = s->d;
  s->d = ((s->d << 21) | (s->d >> 43)) + mix;
  return mix;
}
